src/array.rs

```diff
@@ -183,6 +187,125 @@ impl<'a, A: Clone> TreeLike for Partition<'a, A> {
     }
 }
 
+/// Convert a tree into a binary tree.
+/// Each node has fanout at most two.
+///
+/// Nodes with at most two children (non-arrays) are preserved in the binary tree.
+/// We call these nodes _normal nodes_.
+///
+/// Nodes with more than two children (arrays) are converted into balanced binary trees.
+///
+/// ## Example
+///
+/// Take a root node with children `a`, `b`, `c`.
+///
+/// ```text
+///    .
+///  / | \
+/// a  b  c
+/// ```
+///
+/// The 3-ary root is converted into two 2-ary nodes. The normal nodes `a`, `b`, `c` stay preserved.
+///
+/// ```text
+///     .
+///    / \
+///   .   c
+///  / \
+/// a   b
+/// ```
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub enum BinaryTree<T> {
+    /// Normal node (with at most two children).
+    ///
+    /// Must be constructed via [`BinaryTree::from_tree()`].
```

Comment:
In 61d8b2393ea2d0690e11e38c2b10fcd58b9e1efe:

This is fine, but if you want to say "must be constructed via X" you really ought to split the type into `BinaryTreeInner` (a non-public enum) and `BinaryTree` (a public struct with a private `inner` field). Then callers literally cannot construct the type except the way you say so.

Reply:
I left the comment "must be constructed via X" as a note to self, because I violated loop invariants by manually constructing structs. It is even better to enforce loop invariants via the compiler, which I will do. Thanks for pointing this out.